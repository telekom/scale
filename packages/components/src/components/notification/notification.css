/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

:host {
  --width: 100%;
  --width-toast: 25rem;
  --radius: var(--telekom-radius-standard);
  --spacing-y: var(--telekom-spacing-composition-space-06);
  --spacing-y-inner: var(--telekom-spacing-composition-space-04);
  --spacing-x-aside: var(--telekom-spacing-composition-space-14);

  --duration-in: var(--telekom-motion-duration-transition);
  --duration-out: var(--telekom-motion-duration-transition);
  --easing-in: var(--telekom-motion-easing-enter);
  --easing-out: var(--telekom-motion-easing-exit);
  --translate-x: 0;
  --translate-y: 0;
  --translate-z: 0;

  display: contents;
}

[part~='base'] {
  position: relative;
  display: none;
  box-sizing: border-box;
  border-radius: var(--radius);
  padding-right: var(--spacing-x-aside);
  width: var(--width);
  background: var(--_background-subtle);
  box-shadow: var(--_shadow);
  animation-duration: var(--duration);
}

[part~='base'][part~='open'] {
  display: flex;
}

@keyframes toggle {
  from {
    opacity: 0;
    transform: translate3d(
      var(--translate-x),
      var(--translate-y),
      var(--translate-z)
    );
  }
}

[part~='in'] {
  animation-name: toggle;
  animation-duration: var(--duration-in);
  animation-timing-function: var(--easing-in);
}

[part~='out'] {
  animation-name: toggle;
  animation-direction: reverse;
  animation-duration: var(--duration-out);
  animation-timing-function: var(--easing-out);
}

[part~='variant-informational'] {
  --_background-subtle: var(--telekom-color-functional-informational-subtle);
  --_background-accent: var(--telekom-color-functional-informational-standard);
  --_color-accent: var(--telekom-color-text-and-icon-functional-informational);
}
[part~='variant-warning'] {
  --_background-subtle: var(--telekom-color-functional-warning-subtle);
  --_background-accent: var(--telekom-color-functional-warning-standard);
  --_color-accent: var(--telekom-color-text-and-icon-functional-warning);
}
[part~='variant-success'] {
  --_background-subtle: var(--telekom-color-functional-success-subtle);
  --_background-accent: var(--telekom-color-functional-success-standard);
  --_color-accent: var(--telekom-color-text-and-icon-functional-success);
}
[part~='variant-danger'] {
  --_background-subtle: var(--telekom-color-functional-danger-subtle);
  --_background-accent: var(--telekom-color-functional-danger-standard);
  --_color-accent: var(--telekom-color-text-and-icon-functional-danger);
}

[part~='type-banner'] {
  --_shadow: var(--telekom-shadow-floating-standard);
}
[part~='type-toast'] {
  --translate-x: var(--telekom-spacing-composition-space-05);
  --_shadow: var(--telekom-shadow-top);
  width: var(--width-toast);
}

[part='icon'] {
  display: flex;
  flex-shrink: 0;
  justify-content: center;
  align-items: flex-start;
  width: var(--spacing-x-aside);
  padding-top: var(--spacing-y);
  padding-bottom: var(--spacing-y);
  border-top-left-radius: var(--radius);
  border-bottom-left-radius: var(--radius);
  color: var(--_color-accent);
}
[part~='type-toast'] [part='icon'] {
  background: var(--_background-accent);
  color: var(--telekom-color-text-and-icon-white-standard);
  align-items: center;
}
[part~='type-toast'][part~='variant-warning'] [part='icon'] {
  color: var(--telekom-color-text-and-icon-black-standard);
}

[part='body'] {
  margin-top: var(--spacing-y);
  margin-bottom: var(--spacing-y);
}
[part~='type-toast'] [part='body'] {
  padding-left: var(--spacing-y);
}

[part='heading'] {
  font: var(--telekom-text-style-heading-6);
  line-height: var(--telekom-typography-line-spacing-tight);
}

[part='text'] {
  margin-top: var(--spacing-y-inner);
}

::slotted(*) {
  font: var(--telekom-text-style-body);
}
::slotted(p) {
  margin: 0;
  margin-top: var(--spacing-y-inner);
}

[part='close-button'] {
  --color-ghost: var(--telekom-text-and-icon-standard);
  position: absolute;
  right: var(--telekom-spacing-composition-space-03);
  top: var(--telekom-spacing-composition-space-03);
}

@media screen and (forced-colors: active), (-ms-high-contrast: active) {
  [part~='base'] {
    border: 1px solid white;
  }

  [part='close-button'] {
    --color-ghost: white;
  }
}
