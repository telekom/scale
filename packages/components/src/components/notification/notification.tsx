/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import {
  Component,
  Prop,
  h,
  Element,
  Event,
  EventEmitter,
  State,
  Host,
  Watch,
} from '@stencil/core';
import cn from 'classnames';
import { animationsFinished } from '../../utils/utils';

const ICON_SIZE = 20;

type CloseEventTrigger = 'CLOSE_BUTTON' | 'ATTRIBUTE' | 'TIMEOUT';

export interface BeforeCloseEventDetail {
  trigger: CloseEventTrigger;
}

const iconVariantNameMap = {
  informational: 'scale-icon-alert-information',
  warning: 'scale-icon-alert-warning',
  success: 'scale-icon-action-success',
  danger: 'scale-icon-alert-error',
};

@Component({
  tag: 'scale-notification',
  styleUrl: './notification.css',
  shadow: true,
})
export class Notification {
  @Element() hostElement: HTMLElement;

  /** Heading */
  @Prop() heading: string;
  /** (optional) Type */
  @Prop() type?: 'inline' | 'banner' | 'toast' = 'inline';
  /** (optional) Variant */
  @Prop() variant?: 'danger' | 'warning' | 'success' | 'informational' =
    'informational';
  /** (optional) Visible */
  @Prop({ reflect: true, mutable: true }) opened?: boolean;
  /** (optional) Show the close button */
  @Prop() dismissible?: boolean = false;
  /** (optional) Time in milliseconds until it closes by itself */
  @Prop() delay?: number;
  /** (optional) `aria-live` of element */
  @Prop() innerAriaLive?: string = 'assertive';
  /** (optional) Label for close button */
  @Prop() closeButtonLabel?: string = 'Close';
  /** (optional) `title` for close button */
  @Prop() closeButtonTitle?: string = 'Close';
  /** Default aria-level for heading */
  @Prop() headingLevel: number = 2;
  /** (optional) string prepended to the heading */
  @Prop() ariaHeading?: string = 'Information';
  /** (optional) Injected styles */
  @Prop() styles?: string;

  /** What actually triggers opening/closing the notification */
  @State() isOpen: boolean = this.opened || false;
  @State() animationState: 'in' | 'out' | undefined;
  @State() role: string = 'alert';
  @State() hasTextSlot: boolean = false;
  // @State() hasActionSlot: boolean = false; // unused for now

  /** Fires after the notification has been opened  */
  @Event({ eventName: 'scale-open' }) scaleOpen: EventEmitter<void>;
  /** Fires on every close attempt. Calling `event.preventDefault()` will prevent the modal from closing */
  @Event({ eventName: 'scale-before-close' })
  scaleBeforeClose: EventEmitter<BeforeCloseEventDetail>;
  /** Fires after the notification has been closed */
  @Event({ eventName: 'scale-close' }) scaleClose: EventEmitter<void>;

  private lastCloseEventTrigger: CloseEventTrigger | null = null;

  connectedCallback() {
    if (this.hostElement.hasAttribute('opened')) {
      // Do not use `role="alert"` if opened/visible on page load
      this.role = undefined;
      this.isOpen = true;
    }
    if (this.delay !== undefined) {
      setTimeout(this.timeout, this.delay);
    }
    this.hasTextSlot = this.hostElement.querySelector('[slot="text"]') != null;
    // this.hasActionSlot =
    //   this.hostElement.querySelector('[slot="action"]') != null;
  }

  @Watch('opened')
  openedChanged(newValue) {
    if (newValue === true) {
      this.open();
      this.lastCloseEventTrigger = 'ATTRIBUTE';
    } else if (this.isOpen) {
      this.close();
    }
  }

  open = () => {
    this.isOpen = true;
    this.role = 'alert';
    this.animationState = 'in';
    requestAnimationFrame(async () => {
      await animationsFinished(this.hostElement.shadowRoot);
      this.animationState = undefined;
      this.scaleOpen.emit();
      if (this.delay !== undefined) {
        setTimeout(this.timeout, this.delay);
      }
    });
  };

  close = () => {
    const event = this.scaleBeforeClose.emit({
      trigger: this.lastCloseEventTrigger,
    });
    this.lastCloseEventTrigger = null;
    const prevented = event.defaultPrevented;
    if (prevented) {
      this.opened = true;
      return;
    }
    this.animationState = 'out';
    requestAnimationFrame(async () => {
      await animationsFinished(this.hostElement.shadowRoot);
      this.animationState = undefined;
      this.isOpen = false;
      this.scaleClose.emit();
    });
  };

  timeout = () => {
    this.lastCloseEventTrigger = 'TIMEOUT';
    this.opened = false;
  };

  render() {
    const IconTag = iconVariantNameMap[this.variant];
    return (
      <Host>
        {this.styles && <style>{this.styles}</style>}
        <div
          part={cn(
            'base',
            this.animationState,
            `type-${this.type}`,
            `variant-${this.variant}`,
            this.isOpen && 'open'
          )}
          role={this.role}
        >
          <div part="icon" aria-hidden="true">
            <slot name="icon">
              <IconTag size={ICON_SIZE} selected={this.type === 'toast'} />
            </slot>
          </div>
          <div
            part="body"
            aria-live={this.role === undefined ? undefined : this.innerAriaLive}
          >
            <div
              part="heading"
              role="heading"
              aria-level={this.headingLevel}
              aria-label={`${this.ariaHeading} ${this.heading}`}
            >
              <span>{this.heading}</span>
            </div>
            {this.hasTextSlot && (
              <div part="text">
                <slot name="text"></slot>
              </div>
            )}
          </div>
          {this.dismissible && (
            <scale-button
              part="close-button"
              variant="ghost"
              onClick={() => {
                this.lastCloseEventTrigger = 'CLOSE_BUTTON';
                this.opened = false;
              }}
            >
              <slot name="close-icon">
                <scale-icon-action-circle-close
                  aria-label={this.closeButtonLabel}
                  accessibilityTitle={this.closeButtonTitle}
                  decorative
                  size={ICON_SIZE}
                />
              </slot>
            </scale-button>
          )}
        </div>
      </Host>
    );
  }
}
