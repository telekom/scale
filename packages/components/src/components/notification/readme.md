# scale-notification



<!-- Auto Generated Below -->


## Properties

| Property           | Attribute            | Description                                                                                       | Type                                                    | Default           |
| ------------------ | -------------------- | ------------------------------------------------------------------------------------------------- | ------------------------------------------------------- | ----------------- |
| `ariaHeading`      | `aria-heading`       | (optional) string prepended to the heading                                                        | `string`                                                | `'Information'`   |
| `closeButtonLabel` | `close-button-label` | (optional) Label for close button                                                                 | `string`                                                | `'Close'`         |
| `closeButtonTitle` | `close-button-title` | (optional) `title` for close button                                                               | `string`                                                | `'Close'`         |
| `delay`            | `delay`              | (optional) Time in milliseconds until it closes by itself                                         | `number`                                                | `undefined`       |
| `dismissible`      | `dismissible`        | (optional) Show the close button                                                                  | `boolean`                                               | `false`           |
| `heading`          | `heading`            | Heading                                                                                           | `string`                                                | `undefined`       |
| `headingLevel`     | `heading-level`      | Default aria-level for heading                                                                    | `number`                                                | `2`               |
| `innerAriaLive`    | `inner-aria-live`    | <span style="color:red">**[DEPRECATED]**</span> - ariaRole should replace innerAriaLive<br/><br/> | `string`                                                | `'assertive'`     |
| `innerRole`        | `inner-role`         | (optional) string prepended to the heading                                                        | `"alert" \| "status"`                                   | `'alert'`         |
| `opened`           | `opened`             | (optional) Visible                                                                                | `boolean`                                               | `undefined`       |
| `styles`           | `styles`             | (optional) Injected styles                                                                        | `string`                                                | `undefined`       |
| `type`             | `type`               | (optional) Type                                                                                   | `"banner" \| "inline" \| "toast"`                       | `'inline'`        |
| `variant`          | `variant`            | (optional) Variant                                                                                | `"danger" \| "informational" \| "success" \| "warning"` | `'informational'` |


## Events

| Event                | Description                                                                                        | Type                                  |
| -------------------- | -------------------------------------------------------------------------------------------------- | ------------------------------------- |
| `scale-before-close` | Fires on every close attempt. Calling `event.preventDefault()` will prevent the modal from closing | `CustomEvent<BeforeCloseEventDetail>` |
| `scale-close`        | Fires after the notification has been closed                                                       | `CustomEvent<void>`                   |
| `scale-open`         | Fires after the notification has been opened                                                       | `CustomEvent<void>`                   |


## Shadow Parts

| Part             | Description |
| ---------------- | ----------- |
| `"body"`         |             |
| `"close-button"` |             |
| `"heading"`      |             |
| `"icon"`         |             |
| `"text"`         |             |


## Dependencies

### Depends on

- [scale-button](../button)
- [scale-icon-action-circle-close](../icons/action-circle-close)

### Graph
```mermaid
graph TD;
  scale-notification --> scale-button
  scale-notification --> scale-icon-action-circle-close
  style scale-notification fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
