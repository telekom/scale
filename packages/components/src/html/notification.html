<!DOCTYPE html>
<html dir="ltr" lang="en">
  <head>
    <meta charset="utf-8" />
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0"
    />
    <title>Stencil Component Starter</title>

    <script type="module" src="/build/scale-components.esm.js"></script>
    <link rel="stylesheet" href="/build/scale-components.css" />
    <style type="text/css" media="screen, print">
      html {
        margin: 0;
        padding: 0;
      }

      body {
        margin: 0;
        padding: 4rem;
      }
      body > *:not(script) + *:not(script) {
        display: block;
        margin-top: 2rem;
      }
    </style>
  </head>

  <body>
    <scale-notification
      heading="The heading is mandatory, yes"
    ></scale-notification>

    <scale-notification
      heading="The heading is mandatory, this is a banner (danger)"
      type="banner"
      variant="danger"
      opened
      dismissible
    ></scale-notification>

    <scale-notification
      heading="The heading is mandatory, this is a toast (success)"
      type="toast"
      variant="success"
      dismissible
      opened
    >
      <span slot="text"
        >Lorem ipsum dolor sit amet, consetetur tempordede invidunt ut labore et
        dolore.</span
      >
    </scale-notification>

    <script>
      let n = document.querySelector('scale-notification');
      setTimeout(() => {
        n.opened = true;
      }, 1000);

      setTimeout(() => {
        let n2 = Object.assign(document.createElement('scale-notification'), {
          opened: true,
          heading: 'This is madnessâ€¦',
          variant: 'warning',
          delay: 4000,
        });
        document.body.appendChild(n2);
      }, 2000);
    </script>
  </body>
</html>
