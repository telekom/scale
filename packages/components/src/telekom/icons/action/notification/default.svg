<svg width="24" height="24" xmlns="http://www.w3.org/2000/svg"><path d="M22 16.95c-2.35-2.35-2.7-4.65-2.8-5.8L19 9.4c-.25-3.15-2.55-5.6-5.5-6.25V3c0-.85-.65-1.5-1.5-1.5s-1.5.65-1.5 1.5v.15C7.55 3.8 5.25 6.25 5 9.4l-.15 1.75c-.15 1.1-.5 3.4-2.85 5.8v3.55h7.05C9.3 21.9 10.5 23 12 23c1.5 0 2.7-1.1 2.95-2.5H22v-3.55zM20.5 19h-17v-1.45c1.6-1.7 2.6-3.9 2.8-6.25l.15-1.75C6.75 6.65 9.1 4.5 12 4.5s5.25 2.15 5.5 5.05l.15 1.75c.2 2.35 1.2 4.55 2.8 6.25V19h.05z" fill="#000" fill-rule="evenodd"/></svg>