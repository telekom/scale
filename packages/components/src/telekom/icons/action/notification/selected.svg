<svg width="24" height="24" xmlns="http://www.w3.org/2000/svg"><path d="M22 16.95c-2.35-2.35-2.7-4.65-2.8-5.8L19 9.4c-.25-3.15-2.55-5.6-5.5-6.25V3c0-.85-.65-1.5-1.5-1.5s-1.5.65-1.5 1.5v.15C7.55 3.8 5.25 6.25 5 9.4l-.15 1.75c-.15 1.1-.5 3.4-2.85 5.8v3.55h7.05C9.3 21.9 10.5 23 12 23c1.5 0 2.7-1.1 2.95-2.5H22v-3.55z" fill="#000" fill-rule="nonzero"/></svg>