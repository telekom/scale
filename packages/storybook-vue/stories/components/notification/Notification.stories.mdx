import {
  Meta,
  Story,
  ArgsTable,
  Canvas,
  Description,
} from '@storybook/addon-docs';
import { action } from '@storybook/addon-actions';
import ScaleNotification from './ScaleNotification.vue';

<Meta
  title="Components/Notification"
  component={ScaleNotification}
  argTypes={{
    heading: {
      control: {
        type: 'text',
      },
    },
    type: {
      options: ['inline', 'banner', 'toast'],
      control: {
        type: 'select',
      },
    },
    variant: {
      options: ['informational', 'warning', 'success', 'danger'],
      control: {
        type: 'select',
      },
    },
    opened: {
      defaultValue: true,
      control: {
        type: 'boolean'
      }
    },
    dismissible: {
      defaultValue: true,
      control: {
        type: 'boolean'
      }
    },
    delay: {
      control: {
        type: 'number'
      }
    },
    innerAriaLive: {
      control: {
        type: 'text'
      },
      description: 'Deprecated! innerAriaLive has been replaced by innerRole'
    },
    headingLevel: {
      control: {
        type: 'number'
      }
    },    
    ariaHeading:{
      defaultValue: 'Information:',
      control: {
        type: 'text'
      }
    },
    closeButtonLabel: {
      control: {
        type: 'text'
      }
    },
    closeButtonTitle: {
      control: {
        type: 'text'
      }
    },
    'scale-open': {
      description: 'Fires after the notification has been opened',
      table: {
        category: 'Events',
      },
    },
    'scale-before-close': {
      description: 'Fires on every close attempt. Calling `event.preventDefault()` will prevent the modal from closing',
      table: {
        category: 'Events',
      },
    },
    'scale-close': {
      description: 'Fires after the notification has been closed',
      table: {
        category: 'Events',
      },
    }
  }}
/>

export const Template = (args, context) => ({
  props: {
    text: String,
    ...ScaleNotification.props,
  },
  template: `
    <div>
      <scale-notification
        :heading="heading"
        :type="type"
        :variant="variant"
        :opened="opened"
        :dismissible="dismissible"
        :aria-heading="ariaHeading"
        :delay="delay"
        :inner-aria-live="innerAriaLive"
        :close-button-label="closeButtonLabel"
        :close-button-title="closeButtonTitle"
        @scale-open="emitOpen"
        @scale-before-close="emitBeforeClose"
        @scale-close="emitClose"
      >
        <span v-if="text" slot="text">{{ text }}</span>
      </scale-notification>
    </div>
  `,
  methods: {
    emitOpen: action('scale-open'),
    emitBeforeClose: action('scale-before-close'),
    emitClose: action('scale-close'),
  }
})

<div
  style={{
    display: 'inline-flex',
    alignItems: 'center',
    justifyContent: 'space-between',
    width: '100%',
  }}
>
  <h1>Notification</h1>
  <img src="assets/aa.png" alt="Accessible AA" />
</div>

## Standard

<Canvas withSource="none">
  <Story
    name="Standard"
    args={{
      heading: 'Info notification with icon, title, copytext and close',
      text: 'Lorem ipsum dolor sit amet, consetetur tempordede invidunt ut labore et dolore.',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

<ArgsTable story="Standard" />

```html
<scale-notification
  heading="Success notification with icon, title, copytext and close"
  dismissible
  opened
>
  <span slot="text">Lorem ipsum dolor sit amet, consetetur tempordede invidunt ut labore et dolore.</span>
</scale-notification>
```

### Scoped CSS variables

```css
:host {
  --width: 100%;
  --width-toast: 25rem;
  --radius: var(--telekom-radius-standard);
  --spacing-y: var(--telekom-spacing-composition-space-06);
  --spacing-y-inner: var(--telekom-spacing-composition-space-04);
  --spacing-x-aside: var(--telekom-spacing-composition-space-14);

  --duration-in: var(--telekom-motion-duration-transition);
  --duration-out: var(--telekom-motion-duration-transition);
  --easing-in: var(--telekom-motion-easing-enter);
  --easing-out: var(--telekom-motion-easing-exit);
  --translate-x: 0;
  --translate-y: 0;
  --translate-z: 0;
}
```

## Types

### Inline (default)

<Canvas withSource="none">
  <Story name="Inline" args={{
    heading: 'Info notification with icon, title and close',
  }}>
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-notification
  heading="Info notification with icon, title and close"
  opened
></scale-notification>
```

### Banner

<Canvas withSource="none">
  <Story name="Banner" args={{
    heading: 'Success notification with icon, title and close',
    type: 'banner',
    variant: 'success',
    ariaHeading: 'Success:'
  }}>
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-notification
  heading="Success notification with icon, title and close"
  type="banner"
  variant="success"
></scale-notification>
```

### Toast

<Canvas withSource="none">
  <Story name="Toast" args={{
    heading: 'Warning toast headline',
    type: 'toast',
    variant: 'warning',
    text: 'Lorem ipsum dolor sit amet, consetetur tempordede invidunt ut labore et dolore.',
    ariaHeading: 'Warning:'
  }}>
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-notification
  heading="Warning toast headline"
  type="toast"
  variant="warning"
  dismissible
  opened
>
  <span slot="text">Lorem ipsum dolor sit amet, consetetur tempordede invidunt ut labore et dolore.</span>
</scale-notification>
```

#### Toast Stack

In order to "toast" a notification, append it to a div with the `.scl-toast-stack` class. Listen to the `scale-close` event if you'd like to remove them from the DOM after being dismissed.

<Canvas withSource="none">
  <Story name="Toast Stack">
    {{
      data() {
        return {
          count: 1,
          delay: 0,
          variants: ['informational', 'warning', 'success', 'danger'],
          currentVariant: 0,
        }
      },
      computed: {
        variant() {
          const s = this.variants[this.currentVariant]
          return s.charAt(0).toUpperCase() + s.slice(1)
        }
      },
      template: `
        <div>
          <div ref="stack" class="scl-toast-stack"></div>
          <div style="display: flex; gap: var(--telekom-spacing-unit-x12); align-items: center">
            <scale-button variant="secondary" @click="create">Show</scale-button>
            <scale-slider
              @scale-change="$event => this.delay = $event.target.value"
              label="Delay"
              min="0"
              max="10"
              show-step-marks
              unit="s"
              helper-text="0 means no timeout."
            ></scale-slider>
          </div>
        </div>
      `,
      methods: {
        create() {
          const toast = Object.assign(
            document.createElement('scale-notification'),
            {
              heading: `${this.variant} toast headline #${this.count++}`, 
              type: 'toast',
              variant: this.variants[this.currentVariant],
              delay: this.delay === 0 ? undefined : this.delay * 1000,
              dismissible: true,
              opened: false,
              innerHTML: '<span slot="text">Lorem ipsum dolor sit amet, consetetur tempordede invidunt ut labore et dolore.</span>'
            }
          );
          const handleOpen = () => {
            this.emitOpen();
            toast.removeEventListener('scale-open', handleOpen);
          }
          const handleBeforeClose = (event) => {
            this.emitBeforeClose();
            toast.removeEventListener('scale-before-close', handleBeforeClose);
          }
          const handleClose = () => {
            this.emitClose();
            toast.removeEventListener('scale-close', handleClose);
            toast.remove();
          }
          toast.addEventListener('scale-open', handleOpen);
          toast.addEventListener('scale-before-close', handleBeforeClose);
          toast.addEventListener('scale-close', handleClose);
          this.$refs.stack.prepend(toast);
          toast.opened = true;
          this.currentVariant = (this.currentVariant + 1) % this.variants.length;
        },
        emitOpen: action('scale-open'),
        emitClose: action('scale-close'),
        emitBeforeClose: action('scale-before-close'),
      }
    }}
  </Story>
</Canvas>

```html
<div class="scl-toast-stack">
  <scale-notification
    heading="Informational toast headline"
    type="toast"
    dismissible
    delay="10000"
    opened
  >
    <span slot="text">Lorem ipsum dolor sit amet, consetetur tempordede invidunt ut labore et dolore.</span>
  </scale-notification>
</div>
```

## Variants

### Success

<Canvas withSource="none">
  <Story name="Success" args={{
    heading: 'Success notification with icon and title',
    variant: 'success',
    dismissible: false,
    ariaHeading: 'Success:'
  }}>
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-notification
  heading="Success notification with icon and title"
  variant="success"
  opened
></scale-notification>
```

### Informational (default)

<Canvas withSource="none">
  <Story name="Informational" args={{
    heading: 'Informational notification with icon and title',
    variant: 'informational',
    dismissible: false,
  }}>
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-notification
  heading="Informational notification with icon and title"
  variant="informational"
  opened
></scale-notification>
```

### Danger

<Canvas withSource="none">
  <Story name="Danger" args={{
    heading: 'Danger notification with icon and title',
    variant: 'danger',
    dismissible: false,
    ariaHeading: 'Danger:'
  }}>
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-notification
  heading="Danger notification with icon and title"
  variant="danger"
  opened
></scale-notification>
```

### Warning

<Canvas withSource="none">
  <Story name="Warning" args={{
    heading: 'Warning notification with icon and title',
    variant: 'warning',
    dismissible: false,
    ariaHeading: 'Warning:'
  }}>
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-notification
  heading="Warning notification with icon and title"
  variant="warning"
  opened
></scale-notification>
```
