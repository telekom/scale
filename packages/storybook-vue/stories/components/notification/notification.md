<div style="display: inline-flex; align-items: center; justify-content: space-between; width: 100%;">
    <h1>Notification</h1>
    <img src="assets/beta.png" alt="Beta Component" />
</div>

Notifications give users feedback about the status of the system or an action.

## General

Scale offers toast, banner, and inline notifications.

### When to use

Use notifications to:

- inform users about the status of an action,
- provide a system message, or
- communicate updates like "New Message".

### When not to use

Don't use a notification when:

- the user needs to confirm a decision to prevent or correct critical errors (here, use a modal instead).

## Basic types

### Toast notification

Toast notifications appear immediately after a specific event. They donâ€™t refer to an object on the page and are temporarily displayed in a global position.
Toast notifications can cover other content and hide automatically after a short time. Alternatively, users can actively close them. If necessary, use a link to
provide further information.

![Image Name](assets/3_components/notifications/toast.png)

### Banner notification

Banner notifications communicate system messages. They are displayed above the page content.

![Image Name](assets/3_components/notifications/banner.png)

### Inline notification

Inline notifications give users feedback on an action. They are located in the corresponding section of the page.

![Image Name](assets/3_components/notifications/inline.png)

## Variants

### Informational notification

The informational notification gives users additional helpful information.

![Image Name](assets/3_components/notifications/variant_info.png)

### Success notification

The success notification confirms a successfully performed action.

![Image Name](assets/3_components/notifications/variant_success.png)

### Warning notification

The warning notification informs users about potentially unwanted results.

![Image Name](assets/3_components/notifications/variant_warning.png)

### Error notification

The error notification informs users about a failed action or error and may require an action to continue.

![Image Name](assets/3_components/notifications/variant_error.png)

## Elements

![Image Name](assets/3_components/notifications/elements.png)

### Icon (1)

The icon illustrates the basic message of the notification.

### Headline (2)

Formulate the text in a meaningful way and keep it as short as possible.

### Subheadline (3) (optional)

A supporting text is only necessary in exceptional cases.

### Link (4) (optional)

Use the link within the toast or banner notification (e.g., for further information).

### Close icon (5) (optional)

The close icon is optional.

## Positioning

### Toast notification

Toast notifications are usually placed on the top left or top right and can cover other content.

![Image Name](assets/3_components/notifications/position_toast.png)

### Banner notification

Center banner notifications above the page content.

![Image Name](assets/3_components/notifications/position_banner.png)

### Inline notification

Position inline notifications in the appropriate page area.

![Image Name](assets/3_components/notifications/position_inline.png)

## Beta component

This component is still in the beta phase. When testing it, keep in mind that it might not have gone through all quality control checks, and it might not yet have WCAG accessibility certification. This component might change in the future.
