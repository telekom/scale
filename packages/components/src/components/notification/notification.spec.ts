/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { newSpecPage } from '@stencil/core/testing';
import { Notification } from './notification';

describe('Notification ', () => {
  let element;

  beforeEach(async () => {
    element = new Notification();
  });

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [Notification],
      html: `<scale-notification-toast opened></scale-notification-toast>`,
    });
    expect(page.root).toMatchSnapshot();
  });
  it('should match snapshot, type=banner', async () => {
    const page = await newSpecPage({
      components: [Notification],
      html: `<scale-notification-toast opened type="banner"></scale-notification-toast>`,
    });
    expect(page.root).toMatchSnapshot();
  });
  it('should match snapshot, type=toast', async () => {
    const page = await newSpecPage({
      components: [Notification],
      html: `<scale-notification-toast opened type="toast"></scale-notification-toast>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
