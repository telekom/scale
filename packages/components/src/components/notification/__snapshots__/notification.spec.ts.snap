// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification  should match snapshot 1`] = `
<scale-notification-toast
  opened=""
/>
`;

exports[`Notification  should match snapshot, type=banner 1`] = `
<scale-notification-toast
  opened=""
  type="banner"
/>
`;

exports[`Notification  should match snapshot, type=toast 1`] = `
<scale-notification-toast
  opened=""
  type="toast"
/>
`;
